tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/BlockingMiddleware.cs

```diff
@@ -0,0 +1,116 @@
+﻿// <copyright file="BlockingMiddleware.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+#if !NETFRAMEWORK
+using System.Threading.Tasks;
+using Datadog.Trace.AppSec;
+using Datadog.Trace.AppSec.Transports;
+using Microsoft.AspNetCore.Http;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AspNetCore;
+
+internal class BlockingMiddleware
+{
+    private readonly bool _runSecurity;
+    private readonly bool _endPipeline;
+    private readonly RequestDelegate? _next;
+
+    internal BlockingMiddleware(RequestDelegate? next = null, bool runSecurity = false, bool endPipeline = false)
+    {
+        _next = next;
+        _runSecurity = runSecurity;
+        _endPipeline = endPipeline;
+    }
+
+    private static Task WriteResponse(HttpContext context, SecuritySettings settings, out bool endedResponse)
+    {
+        var httpResponse = context.Response;
+        if (!httpResponse.HasStarted)
+        {
+            httpResponse.Clear();
+            foreach (var cookie in context.Request.Cookies)
+            {
+                httpResponse.Cookies.Delete(cookie.Key);
+            }
+
+            httpResponse.Headers.Clear();
+            httpResponse.StatusCode = 403;
+            var template = settings.BlockedJsonTemplate;
+            if (context.Request.Headers["Accept"].ToString().Contains("text/html"))
+            {
+                httpResponse.ContentType = "text/html";
+                template = settings.BlockedHtmlTemplate;
+            }
+            else
+            {
+                httpResponse.ContentType = "application/json";
+            }
+
+            endedResponse = true;
+            return httpResponse.WriteAsync(template);
+        }
+
+        context.Response.Body.Dispose();
+        endedResponse = true;
+        return Task.CompletedTask;
+    }
+
+    internal async Task Invoke(HttpContext context)
+    {
+        var security = Security.Instance;
+        var endedResponse = false;
+        if (security.Settings.Enabled)
+        {
+            var securityTransport = new SecurityTransport(security, context, Tracer.Instance.ActiveScope.Span as Span);
+            if (_runSecurity)
+            {
+                if (_endPipeline && !context.Response.HasStarted)
+                {
+                    context.Response.StatusCode = 404;
+                }
```

Comment:
Just so I understand, this is to emulate the existing 404 behaviour of the application builder right?

Reply:
yes we set it here, knowing we're the last middleware in the pipeline. 
 as we can't insert easily in a performant way, anything behind the 404 aspnet core middleware, as the `Build` method looks like this:
```

public RequestDelegate Build()
        {
            RequestDelegate app = context =>
            {
                // If we reach the end of the pipeline, but we have an endpoint, then something unexpected has happened.
                // This could happen if user code sets an endpoint, but they forgot to add the UseEndpoint middleware.
                var endpoint = context.GetEndpoint();
                var endpointRequestDelegate = endpoint?.RequestDelegate;
                if (endpointRequestDelegate != null)
                {
                    var message =
                        $"The request reached the end of the pipeline without executing the endpoint: '{endpoint.DisplayName}'. " +
                        $"Please register the EndpointMiddleware using '{nameof(IApplicationBuilder)}.UseEndpoints(...)' if using " +
                        $"routing.";
                    throw new InvalidOperationException(message);
                }

                context.Response.StatusCode = 404;
                return Task.CompletedTask;
            };

            foreach (var component in _components.Reverse())
            {
                app = component(app);
            }

            return app;
        }
```
we could intercept the method end and do the same thing, reversing all components again but it wouldnt be ideal I guess 🤔

Reply:
As this will still be executed, I'm not sure if it's worth doing it twice.. 🤔